ssz_serialization/codec.nim

```diff
@@ -299,10 +320,32 @@ proc readSszValue*[T](input: openArray[byte],
       when supportsBulkCopy(type val[0]):
         if val.len > 0:
           copyMem addr val[0], unsafeAddr input[0], input.len
+
+        # There's no selective invalidation here, because it would require a
+        # potential performance tradeoff, either interfering with bulk copy,
+        # or involving more verification of changed hash entries.
+        if not hashListOperations.isNil:
+          hashListOperations.resetCache()
       else:
-        for i in 0 ..< val.len:
-          let offset = i * elemSize
-          readSszValue(input.toOpenArray(offset, offset + elemSize - 1), val[i])
+        if hashListOperations.isNil:
+          for i in 0 ..< val.len:
+            let offset = i * elemSize
+            readSszValue(input.toOpenArray(offset, offset + elemSize - 1), val[i])
+        else:
+          var prevValue: E
+          hashListOperations.resizeHashes()
+
+          for i in 0 ..< val.len:
+            let offset = i * elemSize
+            assign(prevValue, val[i])
+            readSszValue(input.toOpenArray(offset, offset + elemSize - 1), val[i])
```

Comment:
The general issue with avoiding code duplication here is that the "core" functionality is just one or two lines of code. The `template`, etc versions weren't obvious wins.